--env: support more shells, allow explicit shell selection

gregory-nisbet opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

`brew --env` can be given an explicit shell argument via shell.
The basename of the path is compared against a known list of
shells.

An explicit argument of `--shell=` or `--shell ''`
with an empty string forces brew --env to dump the environment in
the key: value format that it uses when it detects that
stdout is a tty.

If an explicit shell is not supplied from the command line and
stdout is not a tty, then `brew --env` will check the parent process,
then its parent process etc until it finds a process with an
argv[0] that is either a known shell name or a known shell name with
`-` prepended

When I ran the test suite I encountered one failure in an integration test
if I run the test suite under `tcsh` but none if I run it under `bash`. The integration test
will probably need to be updated to use an explicit shell.

Here's an excerpt of the failing case.

```
  1) Failure:
IntegrationCommandTests#test_env [/Users/greg/opt/brew/brew-dev/Library/Homebrew/test/test_integration_cmds.rb:113]:
```

gregory-nisbet commented:
I used the construction

    for x in 1..MAX_ITERATIONS
        # body does not include x
    end

which was rejected by the style-checker / static analysis tool.

Is there a way to run the style tests locally?
Is there anything like `Integer::times` but with early termination?

apjanke commented:
It probably wants you to do
```
(1..MAX_ITERATIONS).each do
   # ... body, and I think you can use 'break' for early exit ...
end

> Is there a way to run the style tests locally?

Yep. `brew style <filename.rb>`. This uses [RuboCop](https://github.com/bbatsov/rubocop), with some Homebrew-specific options.

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
 require "extend/ENV"
 require "build_environment"
 
+
```
Please don't make whitespace only style change like the above two, thanks.

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+  )
+  DEFAULT_SHELL = "sh"
+
+
```
Only need one newline here (and elsewhere). Please follow the existing style used in the file.

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+    # basename and strip leading '-' in login shell
+    name = File.basename path
+
+    if VALID_SHELLS.include? name
```
`return name if VALID_SHELLS.include? name`

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+        return stripped_name
+      end
+    end
+    return nil
```
Don't need `return` as the last statement of a function in Ruby and `nil` is implicit if you haven't already passed something.

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+
+  def get_explicit_shell
+    # parse shell from command line arguments
+    split_arg_index = ARGV.index "--shell"
```
Use `ARGV.value` rather than parsing.

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+    # that looks like a shell
+    candidate_pid = Process.ppid.to_s
+
+    for _ in 1..MAX_ITERATIONS
```
This feels a little smelly/overkill to me. I think it's reasonable to just rely on `$SHELL` if it's not passed explicitly.

apjanke replied:
I disagree. (It was me who asked OP to do it this way instead of using `$SHELL` in the first place.) We know there are some users who switch between shells at run time, including me. And especially if they run in to problems: we're likely to ask them to try things out in `bash` instead of their main shell.

And just from experience working with Oh My Zsh, `$SHELL` is often "wrong" about what shell is currently running. It's like `$EDITOR`: tells you what to run, but not what is running. If we're going to detect at all, we should do it this way.

If including this much code is a problem, then let's fall back to defaulting to bash/Bourne style, and requiring an explicit `--shell` for other shells, but not use `$SHELL`.

UniqMartin replied:
Also using a `for` loop is highly unidiomatic Ruby. In this case, a simple

```ruby
MAX_ITERATIONS.times do
  # Loop body.
end
```

will do (and also reads much nicer).

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+    env_keys = build_env_keys(ENV)
+
+    case shell
+    when "", nil
```
Feels like it'd be nicer to ensure this is never an empty string.

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
       dump_build_env ENV
+    when "fish"
+      env_keys.each do |key|
+        puts "set -gx #{key} #{ENV[key]}"
```
Probably worth pulling these into separate functions that are passed either `key` or `env_keys`

mikemcquaid commented:
Made a bunch of comments here. I like the idea and where it's going but probably needs a few rounds of style tweaks to get there. Thanks!

> When I ran the test suite I encountered one failure in an integration test
> if I run the test suite under tcsh but none if I run it under bash. The integration test
> will probably need to be updated to use an explicit shell.

Could you do this? Thanks.

bfontaine commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+      if VALID_SHELLS.include? stripped_name
+        return stripped_name
+      end
+    end
```
You can simplify all these `if`s with something like:
```
def canonicalize_shell(path)
    name = File.basename(path).sub(/^[-]/, "")
    return name if VALID_SHELLS.include? name
end
```

UniqMartin replied:
And use `.strip_prefix("-")` instead of `.sub(/^[-]/, "")`.

bfontaine commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+        return canonical
+      else
+        STDERR.puts "argument #{shell_name} to --shell option is not a valid shell"
+        exit
```
You should exit with a non-zero code if there was an error.

mikemcquaid replied:
You can also use `odie` to exit with the correct code and puts to stderr.

bfontaine commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+        shell = ancestor_shell
+      else
+        shell = DEFAULT_SHELL
+      end
```
Shorter: `shell = get_ancestor_shell || DEFAULT_SHELL`. Also you might want to rename `get_ancestor_shell` to `ancestor_shell`.

bfontaine commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -1,20 +1,128 @@
+      end
+    when "csh", "tcsh"
+      env_keys.each do |key|
+        puts "setenv #{key} #{ENV[key]}"
```
You should probably put quotes around `#{ENV[key]}`.

xu-cheng commented:
Just want to point out:
* You can find a existed implemention to find present shell at here https://github.com/Homebrew/homebrew-command-not-found/blob/master/cmd/brew-command-not-found-init.rb#L3-L17
* I think the method of finding parent/preferred shell as well as how to export value in these shells can be used in multiple place of Homebrew. So I think it is a better idea to put shell related methods to a new file, e.g. `Library/Homebrew/utils/shell.rb`. It could be something like this:
```ruby
module Utils
  module Shell
    def self.preferred_shell
    end

    def self.parent_shell
    end

    def self.shell_profile
      # move current method at https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils.rb#L510-L517 to here
    end

    def self.export_value(shell, key, value)
    end
  end
end
```

gregory-nisbet commented:
Still todo:
1) tests
2) refactor shell-related utility commands to utility class
2a) use symbols for shell names instead of strings
3) ARGV
4) Do we look at the argv[0] of just the parent processes or multiple processes before falling back to `SHELL`?

So, I cleaned up the style a little bit, fixed the failing integration test, and added `sh_quote` (also works for fish) and `csh_quote` implementations that should probably be refactored into the new shell utility class. The existing practice of putting double quotes around the raw value breaks for values containing `"`, `$`, `` ` ... ` ``, `$( ... )` (and possibly more) in sh-derived shells and for values containing `"`, `$`, `!`, `` ` ... ` `` or a literal newline in tcsh.

I'm also overloading `--shell` so that with an an explicit empty value `--shell=`, it prints in the `key: value` and doesn't quote anything. This is the format used when brew detects that the output is a tty. This format is convenient to script against if you need to parse the output for some reason.

@mikemcquaid

I fixed the style issues. As for the parsing `ARGV` I can't seem to find documentation for a `value` method for arrays or for `ARGV` specifically (which I think is just an array). I essentially haven't written any Ruby before so even basic stuff is tricky :/

@xu-cheng  

I just want to get a better understanding of what the different methods in the proposed utility module do.

Does `self.preferred_shell` return the value of the `SHELL` environment variable or does it return the value taken from the command line?

xu-cheng commented:
@gregory-nisbet For example, you can:
```ruby
module Utils
  module Shell
    def self.preferred_shell
      case ENV["SHELL"]
      when %r{/(ba)?sh} then :bash
      when %r{/zsh} then :zsh
      when ... then ...
      else :bash # default to bash
      end
    end

    def self.parent_shell
      case `ps c -p #{Process.ppid} -o 'ucomm='`.chuzzle
      when %r{/(ba)?sh} then :bash
      when %r{/zsh} then :zsh
      when ... then ...
      else preferred_shell
      end
    end

    def self.shell_profile
      # move current method at https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils.rb#L510-L517 to here
     case preferred_shell
     when :bash then "~/.bash_profile"
     ...
     end
    end

    def self.export_value(shell, key, value)
      case shell
      when :bash, :zsh then "export #{key}=\"#{value}\""
      ...
      end
    end
  end
end
```

bfontaine commented:
@gregory-nisbet You might want to look at the [`shellwords`](http://ruby-doc.org/stdlib-2.3.0/libdoc/shellwords/rdoc/Shellwords.html) module for the quotes thing.

mikemcquaid commented:
> I fixed the style issues. As for the parsing ARGV I can't seem to find documentation for a value method for arrays or for ARGV specifically (which I think is just an array). I essentially haven't written any Ruby before so even basic stuff is tricky :/

Don't worry, we're happy to help. `ARGV.value` is a Homebrew internal thing; you can see a usage example here: https://github.com/Homebrew/brew/blob/1fa48234e5301f018dc93e62bc7e03d71729b296/Library/Homebrew/cleanup.rb#L141-L149

gregory-nisbet commented:
I added the `utils/shell.rb` file and some tests to `test_intregation_cmds.rb` .

The `shell_profile` method (https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils.rb#L510-L517) has been moved and its tests updated.

Broadly speaking, most of the changes I made involving `shell_profile` are probably wrong / not idiomatic since I used `Utils::Shell.shell_profile` all over the place. I'm still trying to figure out the module system works and how pseudo-static methods (`def self.method_name`) work.

`ARGV.value` does not seem to distinguish an explicit value of empty string, e.g.

When executing this command

    $ brew --env --shell=

The value of `ARGV.value("shell")` would be `nil`.

I think it should be empty string, but other parts of the codebase may depend on this behavior. So right now I'm using `--shell=none` to explicit force `brew --env` to output the build environment in the `key: value` format. I'm thinking it should probably either be `--shell=` with an empty argument or a dedicated command line option like `--noshell` or `--prettyprint` or something.

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -3,19 +3,37 @@
+    # then we check what kind of shell the parent process is
+    # finally we consult the environment variable
+    shell = explicit_shell()
+    shell = :noshell if $stdout.tty? and not shell
```
Prefer `||` over `and not`

mikemcquaid replied:
Also you can do `||=` instead of `unless shell` or `and not shell`

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -3,19 +3,37 @@
     else
-      build_env_keys(ENV).each do |key|
-        puts "export #{key}=\"#{ENV[key]}\""
+      if shell then
```
Don't need the `then`

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -3,19 +3,37 @@
+    # then we detect if the output is a tty
+    # then we check what kind of shell the parent process is
+    # finally we consult the environment variable
+    shell = explicit_shell()
```
Don't need to append `()`

mikemcquaid replied:
(for functions in general)

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -3,19 +3,37 @@
+  def explicit_shell
+    shell = ARGV.value("shell")
+    return :noshell if shell == "none"
+    return Utils::Shell.path_to_shell(shell) if shell
```
The last `return` is unnecessary.

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -3,19 +3,37 @@
 module Homebrew
+  def explicit_shell
+    shell = ARGV.value("shell")
+    return :noshell if shell == "none"
```
Can this not just be a `nil` value?

gregory-nisbet replied:
`explicit_shell` returns `nil` when no shell has been passed in through the command line and `:noshell` when `--shell=none` has been specified. `--shell=none` is probably not the right way to do this but I wanted some way to force the `key: value` format. Originally I was using `--shell=`, but `ARGV.value` returns nil when an explicit empty value is passed and when the option is not passed at all.

UniqMartin replied:
More lines, but (in my opinion) easier to read would be to make this and the following line:

```ruby
if shell == "none"
  :noshell
elsif shell
  Utils::Shell.path_to_shell(shell)
end
```

mikemcquaid replied:
Can you elaborate a bit on what format you want displayed with no `--shell`, `--shell=none` and `--shell=$SHELL`? Thanks!

gregory-nisbet replied:
With no `--shell`, I was going to keep the behavior the same as it is now. If stdout is a tty, `brew --env` outputs in the `key: value` format.

current behavior:

stdout is tty

    $ brew --env
    CMAKE_PREFIX_PATH: /usr/local
    ...
    PATH: /usr/local/Library/ENV/4.3:/usr/bin:/bin:/usr/sbin:/sbin

stdout is not tty

    $ brew --env | cat
    export CC="clang"
    ...
    export ACLOCAL_PATH="/usr/local/share/aclocal"

There is currently no way to force `brew --env` to print the environment variables and their values in the `key: value` format, stdout has to really be a tty. That's why I was thinking of adding something like `--shell=none`.

mikemcquaid replied:
Am I correct saying that `--shell` if your shell is Bash/ZSH or `--shell=bash` will preserve the same behaviour as there is currently? If so, I think that's enough to just require `--shell=bash` rather than a `--shell=none` which implies the same.

gregory-nisbet replied:
Right now `--shell` or `--shell=` is not handled distinctly from the no arguments case (because `ARGV.value("shell")` doesn't distinguish these cases, I think.) and has the original behavior if your shell happens to be `bash`, yes.

`--shell=bash`, however, does not currently preserve the original behavior, since it overrules the interactiveness check.

`--shell=none` is a silly hack to force the `brew --env` to output in `key: value` format. Without `--shell=none` there isn't a way to get the colon format for the environment variables without faking a tty as far as I know. `--shell=none` should probably be renamed.

Just out of curiosity, why does the original behavior of `brew --env` care about whether output is a tty or not? It might make sense to remove that check if `brew --env` is going to take arguments anyway.

UniqMartin replied:
Our `ARGV.value` has a few problems, one of them being that it doesn't accept an empty value after the equals sign (it expects at least one character). I find this surprising and undesirable, thus opened #231 to address this. You could start using `--shell=` if that goes through.

Here's a proposal (both regarding behavior and option names):

- `brew --env` preserves existing behavior, i.e. uses the colon-delimited format for TTY and `bash`-compatible format otherwise. (We want to preserve this behavior as not not break existing scripts that may rely on this.)
- `brew --env --plain` always uses the colon-delimited format (currently `--shell=none`).
- `brew --env --export[=auto]` auto-detects the shell and uses a format suitable for exporting to the detected shell.
- `brew --env --export=<shell>` uses a format suitable for exporting to the explicitly selected shell.

(Or maybe just use `--format=(plain|auto|<shell>)` to cover the latter three cases?)

gregory-nisbet replied:
@UniqMartin, I like that better then what I have now, I'll make the changes. Just a few questions.

- for autodetecting the shell, do we still want to check the argv[0] of the parent process and then fall back to the value of `SHELL`? That seems to be what folks have settled on but I want to make sure.
- now that `--env` is a bona fide subcommand does it make sense to make it `brew env` (and keep the `--env` for backward compatibility).

UniqMartin replied:
> for autodetecting the shell, do we still want to check the argv[0] of the parent process and then fall back to the value of `SHELL`? That seems to be what folks have settled on but I want to make sure.

Yes, that was also my understanding and I'm in support of this approach.

> now that `--env` is a bona fide subcommand does it make sense to make it `brew env` (and keep the `--env` for backward compatibility)?

I agree that passing additional options to that feels a bit weird, but let's stick with this command name for now. We've already renamed it from `environment` to `--env` in the past and I'm pretty hesitant to add yet another alias. (In any case, and if other maintainers feel like you, that's something that should go into a follow-up PR. Let's keep this one more focused, so that it can be merged sooner than later.)

mikemcquaid commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,68 @@
+    def self.path_to_shell(path)
+      # remove leading - from login shell
+      case File.basename(path).strip_prefix("-")
+      when "sh" then :sh
```
Could could combine all these `when`s and use `to_sym` on the string value.

gregory-nisbet commented:
The last few builds seem to be failing at least in part because a merge conflict that I resolved manually (affecting `diagnostics.rb`) in this branch seems to be producing an invalid patch file or something. (Although I think there's also a problem related to ruby 1.8 compatibility ...)

This PR now has a bunch of changes in it that could be factored into two PRs... one for `utils/shell.rb` and one for `csh` and `fish` support.

UniqMartin commented:
I allowed myself to tweak the PR title to make it more descriptive. Feel free to adjust it if I failed to capture the gist of this PR.

> This PR now has a bunch of changes in it that could be factored into two PRs... one for `utils/shell.rb` and one for `csh` and `fish` support.

I think it's fine to keep them together. After all, `utils/shell.rb` grew out of the needs of this PR. However, you could rebase your work on top of `master` (if you haven't already), then squash and possibly amend the commits in your branch so that you end up with these two commits:

1. Introduces `utils/shell.rb`. Makes adjustments to affected parts of Homebrew.
2. Enhances `cmd/--env.rb`.

(This will also help resolve the problem that our test bot is currently unable to pull your changes.)

UniqMartin commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -532,6 +533,19 @@ def check_user_curlrc
+          and remove it from #{Utils::Shell.shell_profile} if present.
+        EOS
+      end
+
```
This block of code probably originates from the merge conflict you resolved manually. It should no longer be here, but instead in `extend/os/mac/diagnostic.rb`.

UniqMartin commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,64 @@
+require 'shellwords'
```
Please always use double quotes for strings. More importantly: The `shellwords` module doesn't exist in Ruby 1.8.7, a version of Ruby we still support, though we're in the process of phasing out that support. This is the main reason why tests on our OS 10.9 test bot fail. I can think of two solutions to this problem:

1. Only use `shellwords` with Ruby 2.0 and above, fall back to not quoting (like before) for older Rubies.
2. Re-implement the required functionality, like you already did with `csh_quote`, and drop the `shellwords` dependency completely (if that's feasible).

bfontaine replied:
My mistake; I suggested `shellwords` without checking if it was available for 1.8.7 :see_no_evil:

UniqMartin commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,64 @@
+      when :tcsh then "~/.tcshrc"
+      when :csh then "~/.cshrc"
+      when :sh then "~/.bash_profile" # /bin/sh is POSIX-mode bash on OS X
+      when :fish then "~/.config/fish/config.fish"
```
It would be nice to have these cases sorted by name. Even better (I think) would be to move the mapping into a constant `Hash` and fetch the entries from it. This could look something like this:

```ruby
SHELL_PROFILE_MAP = {
  :bash => "~/.bash_profile",
  :csh => "~/.cshrc",
  :fish => "~/.config/fish/config.fish",
  :ksh => "~/.kshrc",
  :sh => "~/.bash_profile", # /bin/sh is POSIX-mode bash on OS X
  :tcsh => "~/.tcshrc",
  :zsh => "~/.zshrc",
}.freeze

def self.shell_profile
  SHELL_PROFILE_MAP.fetch(preferred_shell, "~/.bash_profile")
end
```

UniqMartin commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,64 @@
+    # quote values. quoting keys is overkill
+    def self.export_value(shell, key, value)
+      case shell
+      when :bash, :sh, :zsh, :ksh
```
Minor detail, but I'd prefer to keep these sorted alphabetically.

UniqMartin commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,64 @@
+        # a single quote can be included in a single-quoted string via \'
+        # and a literal \ can be included via \\
+        "set -gx #{key} #{Shellwords.shellescape(value)}"
+      when :tcsh, :csh
```
Minor detail, but I'd prefer to keep these sorted alphabetically.

UniqMartin commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,64 @@
+      # remove leading - from login shell
+      shell_name = File.basename(path).strip_prefix("-")
+      case shell_name
+      when "sh", "bash", "zsh", "csh", "tcsh", "fish", "ksh"
```
Minor detail, but I'd prefer to keep these sorted alphabetically.

UniqMartin commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,64 @@
+      when "sh", "bash", "zsh", "csh", "tcsh", "fish", "ksh"
+        shell_name.to_sym()
+      else nil
+      end
```
The `else nil` case seems redundant. You could also replace the entire `case` statement with the following line:

```ruby
shell_name.to_sym if %w[bash csh fish ksh sh tcsh zsh].include?(shell_name)
```

UniqMartin commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,64 @@
+      # we need to prepend a \ to literal newlines and
+      # ! (history substitution character)
+      substrings = str.split("'", -1)
+      substrings.map { |s| "'" + (s.gsub(/([\n!])/) { '\\' + $1 }) + "'" }.join("\"'\"")
```
You can make `s.gsub(/([\n!])/) { '\\' + $1 }` more efficient if you provide it directly with a replacement string instead of passing a block:

```ruby
s.gsub(/[\n!]/, "\\\\\\0")
```

(The many backslashes look somewhat horrible, but they are required because one level of backlash escapes is needed for the string and another level is needed for the replacement placeholders `\<n>`.)

gregory-nisbet commented:
I added some unit tests for `utils/shell.rb` in `tests/test_shell.rb`, alphabetized the shells in `utils/shell.rb`. and updated the unit tests.

I changed the way that `csh_quote` and `sh_quote` work to be more like `Shellwords.shellescape`. Instead of wrapping the whole thing with `"..."`, it just inserts `\` as needed and replaces newline with "\n" or "\\\n" as appropriate.

xu-cheng commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+
+    def self.parent_shell
+      path_to_shell(`ps -p #{Process.ppid} -o command=`.chomp)
```
I think this should be
```ruby
`ps -p #{Process.ppid} -o ucomm=`.strip
```

Noted with `ucomm=`, there won't be leading - from login shell.

xu-cheng commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,71 @@
+    def self.path_to_shell(path)
+      # remove leading - from login shell
+      shell_name = File.basename(path).strip_prefix("-")
+      shell_name.to_sym if %w[bash csh fish ksh sh tcsh zsh].include?(shell_name)
```
This function doesn't seem to handle case that shell path is appended with version. e.g. `zsh-5.3`

xu-cheng commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -513,7 +514,7 @@ def check_user_path_3
           formulae that put executables in #{HOMEBREW_PREFIX}/sbin.
           Consider setting the PATH for example like so
-            echo 'export PATH="#{HOMEBREW_PREFIX}/sbin:$PATH"' >> #{shell_profile}
+            echo 'export PATH="#{HOMEBREW_PREFIX}/sbin:$PATH"' >> #{Utils::Shell.shell_profile}
```
Considering we start to support shell like fish, the `echo 'export ...` part have to handle in a shell by shell case.

gregory-nisbet replied:
I'm really not sure what to do here, since this is a user-facing message. Were you thinking of detecting the shell and sending an appropriate message?

In that case, I think the fish message should show something like this for `fish`.

    fish: 
      1) universal variable.
      set -U fish_user_paths #{HOMEBREW_PREFIX}/sbin $fish_user_paths

      2) modify /.config/fish/fish.config
      echo 'set -x PATH "#{HOMEBREW_PREFIX}"/sbin $PATH' >> #{Utils::Shell.shell_profile}

From the fish tutorial: http://fishshell.com/docs/current/tutorial.html#tut_path

> A faster way is to modify the $fish_user_paths universal variable, which is automatically prepended to $PATH. For example, to permanently add /usr/local/bin to your $PATH, you could write:

`tcsh` would just be 

    tcsh:
        set path = ( #{HOMEBREW_PREFIX}/sbin $path:q )

xu-cheng replied:
It would be still nice to fix this. Maybe by creating a new method like.
```ruby
<<-EOS.undent
 Consider setting the PATH for example like so
   #{Utils::Shell.prepend_path_in_shell_profile(HOMEBREW_PREFIX/"sbin")}
EOS
```

And
```ruby
def prepend_path_in_shell_profile(path)
  case preferred_shell
  when :bash, :zsh
    "echo '#{sh_quote(path)}:$PATH >> #{shell_profile}'
  when ...
  end
end

xu-cheng commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,71 @@
+    # to a shell, return nil if there's no match
+    def self.path_to_shell(path)
+      # remove leading - from login shell
+      shell_name = File.basename(path).strip_prefix("-")
```
as mentioned in https://github.com/Homebrew/brew/pull/201/files#r62985414, `strip_prefix("-")` can be avoid if use different ps command argument.

mikemcquaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -3,20 +3,36 @@
+    # then we check what kind of shell the parent process is
+    # finally we consult the environment variable
+    shell = explicit_shell
+    shell ||= :noshell if $stdout.tty?
```
I think you can make shell `nil` in this case and use that everywhere you should have used `:noshell`.

mikemcquaid commented:
> Also ... how do I squash the commits so the tests can be run on the build bot?

Squash commits by running `brew update; git checkout $YOUR_BRANCH; git rebase --interactive origin/master` and then mark each commit other than the first as `squash` or `fixup` then `git push -f`. Thanks!

gregory-nisbet commented:
Okay, I'm really confused now. In the latest test run, I'm seeing the following statuses for the result of two tests in Yosemite, El Capitan, and Mavericks.

    tests	1 min 18 sec	Fixed
    tests --no-compat	56 sec	Fixed

I don't see any errors, but the build is still marked as failing. I'm also seeing `test (no failures)` in the summary for each of the OSes.

I've incorporated all the feedback I've gotten so far (thanks again for your patience, everyone), except for @xu-cheng's suggestion to change the `export key=value` message to something shell-specific in a few places because I'm not sure what to do.

I followed @UniqMartin's suggestion to change `brew --env` to have backwards-compatible behavior (I think the only thing that I changed is escaping the contents of variables). Meaning that it will use the plain output if stdout is a tty and bash-compatible syntax if it is not. An explicit shell can be specified with `--shell` and a shell will be inferred if `--shell=auto`. The `--env` option also honors `--plain`.

mikemcquaid commented:
> I don't see any errors, but the build is still marked as failing. I'm also seeing test (no failures) in the summary for each of the OSes.

Ignore the failure, it's a false negative.

DomT4 commented:
@BrewTestBot test this please

gregory-nisbet commented:
Wait ... no way. It's passing now? What was up with the test bot? Also, the way the `csh_quote` works is broken because it doesn't insert double quotes when inserting a single-quoted newline. I'm going to update that now and add a test.

DomT4 commented:
> What was up with the test bot?

It was suffering from a mild case of the ol' FUBAR. We're hoping to fix the real issue at some point but for now there's a mitigation in place.

gregory-nisbet commented:
I have modified the help text in diagnostic.rb to be shell-specific and added tests for the new features.

gregory-nisbet commented:
So there might not be enough csh users to justify including that specifically and I can remove it, but I think the Utils::Shell thing and shell-specific profile advice is probably useful. I don't think there are any open style concerns at the moment...

MikeMcQuaid commented on Library/Homebrew/cmd/--env.rb:
```diff
@@ -11,12 +12,25 @@ def __env
 
-    if $stdout.tty?
+    shell_value = ARGV.value("shell")
+    has_plain = ARGV.include?("--plain")
```
I'd inline this variable so it's easier to follow as it's only used once.

MikeMcQuaid commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,84 @@
+
+  module Shell
+    # take a path and heuristically convert it
+    # to a shell, return nil if there's no match
```
`to a shell name`?

MikeMcQuaid commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,84 @@
+      return "''" if str.empty?
+      str = str.dup
+      # anything that isn't a known safe character is padded
+      str.gsub!(/([^A-Za-z0-9_\-.,:\/@\n])/, "\\\\" + "\\1")
```
This looks identical to the method below; maybe worth pulling out into a shared method?

gregn-adroll replied:
`csh_quote` is almost identical except for the way it handles newlines. A newline cannot be escaped in `csh` the same way you would in `sh` so `\n` gets mapped to `'\\\n'`

gregory-nisbet replied:
`csh_quote` is almost the same except for the way it handles newlines. `\n` gets mapped to `'\\\n'` in `csh` to avoid stopping the command prematurely if there's a newline.

MikeMcQuaid replied:
Sorry, I mean the regex on this line looks identical and could be pulled out.

gregory-nisbet replied:
Oh. Yes the initial backslash escaping is shared between sh_quote and csh_quote. I'll fix it when I get home.

MikeMcQuaid commented:
This is looking really good now. A few final, minor tweaks and conflicts that need resolved and I think we can 🚢 this. Thanks for all the patience and hard work @gregory-nisbet: almost there!

xu-cheng commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -498,7 +499,7 @@ def check_user_path_2
           Homebrew's bin was not found in your PATH.
           Consider setting the PATH for example like so
-            echo 'export PATH="#{HOMEBREW_PREFIX}/bin:$PATH"' >> #{shell_profile}
+            #{Utils::Shell.prepend_path_in_shell_profile("#{HOMEBREW_PREFIX}/bin:$PATH")}
```
`:$PATH` should be removed.

xu-cheng commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -513,7 +514,7 @@ def check_user_path_3
           formulae that put executables in #{HOMEBREW_PREFIX}/sbin.
           Consider setting the PATH for example like so
-            echo 'export PATH="#{HOMEBREW_PREFIX}/sbin:$PATH"' >> #{shell_profile}
+            #{Utils::Shell.prepend_path_in_shell_profile("#{HOMEBREW_PREFIX}/sbin:$PATH")}
```
`:$PATH` should be removed.

xu-cheng commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,84 @@
+    def self.prepend_path_in_shell_profile(path)
+      case preferred_shell 
+      when :bash, :ksh, :sh, :zsh, nil
+        "echo 'export PATH=\"#{sh_quote(path)}:$PATH' >> #{shell_profile}"
```
Missing right `\"`, it should be 
```
 "echo 'export PATH=\"#{sh_quote(path)}:$PATH'\" >> #{shell_profile}"
```

xu-cheng commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,84 @@
+      when :csh, :tcsh
+        "echo 'setenv PATH #{csh_quote(path)}:$PATH' >> #{shell_profile}"
+      when :fish
+        "echo 'set -g fish_user_paths $fish_user_paths \"#{sh_quote(path)}\"' >> #{shell_profile}"
```
insert path should before fish_user_paths right? Considering we are perpending path.

DomT4 commented:
Just to check, will `shell_profile` continue working or does every example need to be changed?

We use `shell_profile` in at least a couple of formulae (And it should be used more widely really, rather than assuming everyone in the world uses `bash`).

UniqMartin commented:
Sorry for the late comment and for the long intermediate silence @gregory-nisbet! I'd like to give this a more thorough test and a review, but I'm a bit short on time this weekend. I'd appreciate if merging this could wait a few more days. (I'll try to follow up with a proper review as soon as possible.)

gregory-nisbet commented:
Thanks for your patience and feedback, everyone.

@MikeMcQuaid

I factored out the regex into a constant. I’m not sure what to call it since newlines are also unsafe (in fact, they are so unsafe they need to be quoted individually).

`UNSAFE_SHELL_CHAR = /([^A-Za-z0-9_\-.,:\/@\n])/ `
`sh_quote` and `csh_quote` are still pretty similar, not really sure what to do since the right place for `sh_quote` would probably be in some kind of compatibility wrapper for ruby 1.8 since it’s the same as shell_escape. `csh_quote` could easily be just a call to `shell_escape` that throws an exception if the string contains a newline.

@DomT4
 As for `shell_profile`, the old shell_profile is still there in `Library/Homebrew/utils.rb`, so there shouldn’t be a problem with scripts that depend on it. I can add a deprecation notice to the old definition.

Since the new method is in a new utility module I think it needs to be referred to as `Utils::Shell.shell_profile`, but I’m not sure how the Ruby module system / strategy for namespace management actually works.

@UniqMartin

Sounds good. I look forward to your feedback.

MikeMcQuaid commented:
> I factored out the regex into a constant. I’m not sure what to call it since newlines are also unsafe (in fact, they are so unsafe they need to be quoted individually).

👍 

> As for shell_profile, the old shell_profile is still there in Library/Homebrew/utils.rb, so there shouldn’t be a problem with scripts that depend on it. I can add a deprecation notice to the old definition.

Yeh, add one using `opoo` and move it to `Library/Homebrew/compat/utils.rb`, thanks.

gregory-nisbet commented:
Sorry for the delay. Moved the `shell_profile` function, added deprecation warning.

MikeMcQuaid commented:
LGTM. Any final thoughts, @UniqMartin?

xu-cheng commented on Library/Homebrew/utils/shell.rb:
```diff
@@ -0,0 +1,87 @@
+    def self.prepend_path_in_shell_profile(path)
+      case preferred_shell 
+      when :bash, :ksh, :sh, :zsh, nil
+        "echo 'export PATH=\"#{sh_quote(path)}:$PATH'\" >> #{shell_profile}"
```
There seems to be a misplaced quote here. `'\"` -> `\"'`

gregory-nisbet replied:
Fixed.

MikeMcQuaid commented:
> An explicit argument of --shell= or --shell ''
> with an empty string forces brew --env to dump the environment in
> the key: value format that it uses when it detects that
> stdout is a tty.

`brew --env --shell=''` appears to run as `brew --env` does. `brew --env --shell '' ` produces `Error: No available formula with the name ""`. 

`brew --env --shell=bash` produces `export CXX="clang\+\+"` which seems wrong.

MikeMcQuaid commented:
@gregory-nisbet any news here?

gregory-nisbet commented:
@MikeMcQuaid 

Thanks for the ping and sorry for the long delay.

I think that `brew --env --shell ''` is bypassing `cmd/--env.rb` completely.

As a quick experiment I added a debug `puts` in `cmd/--env.rb` and doesn't print in the `brew --env --shell ''` case, but does with `--shell=''` and `--shell bash`.

```
  8 module Homebrew
  9   def __env
 10     ENV.activate_extensions!
 11     ENV.deps = ARGV.formulae if superenv?
 12     ENV.setup_build_environment
 13     ENV.universal_binary if ARGV.build_universal?
 14
 15     shell_value = ARGV.value("shell")
 16     puts "woof"
 17     puts shell_value
```

as for escaping the `+` when the shell is set to bash, this is what `Shellwords.escape` does.

```
irb(main):010:0> Shellwords.escape '+'
=> "\\+"
```

If it's overkill / looks bad in this case I can change the `sh_quote` function to be less aggressive.

`export CXX="clang\+\+"`

I can add a test for the `brew --env -shell ''` case and try to track down why it isn't delegating to the `--env` subcommand like we would expect.

MikeMcQuaid commented:
> If it's overkill / looks bad in this case I can change the sh_quote function to escape things less aggressively.

That's fine then, ignore me 👍 

> I can add a test for the brew --env --shell '' case and try to track down why it isn't delegating to the --env subcommand like we would expect.

Thanks 👍

gregory-nisbet commented:
regarding the mysterious `brew --env --shell ''` thing. The same error shows up if you give it a name that isn't a valid formula.

`brew --env --shell fishl` for instance does fails with the same error
```Error: No available formula with the name "fishl"```

`fish`, `bash`, `zsh` and the other shells I tested just happened to also be valid formula names.

I'm going to change the `--env` command so that it doesn't keep processing unused arguments as formula names.

gregory-nisbet commented:
Is there a plan for cleaning up argument parsing and making it more consistent?

I think fixing the `brew --env --shell <whatever>` case should be a consequence of cleaning up the argument parsing rather than special-cased, since we're grabbing the command line argument with `ARGV.value` anyway.

`brew --env <name_of_formula>` does have a valid use case. It's just that most formulas happen not to modify the environment.

UniqMartin commented:
I'm afraid I don't understand why the observed behavior is a problem. `ARGV.value("shell")` returns `<whatever>` from `--shell=<whatever>` (note the equals sign instead of a space). It doesn't work if used with `--shell <whatever>`, but it happily returns an empty string for `--shell=` (i.e. an empty value to the `--shell` option). `--shell <whatever>` could be handled by `ARGV.include?("--shell")` followed by `ARGV.next`, but I'm strongly in favor of only supporting the `--shell=` style.

Sorry if I'm completely missing the point and for how long the PR has been sitting here. I've failed to set aside some time to review it and to help get it to a mergeable state …

MikeMcQuaid commented:
@gregory-nisbet @UniqMartin What's your understanding of what needs addressed here before merge?

gregory-nisbet commented:
I don't think there's anything else that needs to be addressed, since some of the quirks of argument parsing have a reasonable explanation.

MikeMcQuaid commented:
Looks good to me locally. Thanks for the hard work and patience here @gregory-nisbet! Without people like you submitting PRs we couldn't run this project. You rock! 👏
